File: C++/4963_섬의개수.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

struct Pos{
	int h, w;
};

int main() {
	int dir[8][2] = {{1,0},{-1,0},{0,1},{0,-1},{1,1},{-1,1},{1,-1},{-1,-1}}; // 방향(상,하,좌,우,대각선 4방향)
	while(1){
		int w, h;
		cin >> w >> h;
		
		if(w == 0 && h == 0){ // while 탈출(종료) 조건
			break;
		}
		
		int **map = new int*[h];
		for(int i = 0; i < h; i++){
			map[i] = new int[w];
		}
		
		vector<Pos> v; // 육지인 좌표를 담을 배열
		for(int i = 0; i < h; i++){
			for(int j = 0; j < w; j++){
				cin >> map[i][j];
				
				if(map[i][j] == 1){
					v.push_back({i,j});
				}
			}
		}
		
		int num = 0; // 영역의 수를 세기 위한 변수
		int v_len = v.size();
		for(int l = 0; l < v_len; l++){ // 육지인 좌표를 root로 하여 bfs 진행
			if(map[v[l].h][v[l].w] == 0){ // 원래는 육지였지만 탐색을 통해 이미 방문한 경우
				continue; // pass하기
			}
			
			num++;
			
			queue<Pos> q; // bfs를 진행하기 위한 queue
			q.push(v[l]);
			map[v[l].h][v[l].w] = 0; // 방문한 경우 중복 이동하지 못하게 0으로 바꾸기
			while(!q.empty()){
				Pos p = q.front();
				q.pop();
				
				for(int i = 0; i < 8; i++){
					int tmp_h, tmp_w;
					tmp_h = p.h + dir[i][0];
					tmp_w = p.w + dir[i][1];
					
					if(tmp_h >= 0 && tmp_h < h && tmp_w >= 0 && tmp_w < w){ // 범위 넘어가지 않게 예외처리
						if(map[tmp_h][tmp_w] != 0){ // 방문한 적 없는 육지일 때
							map[tmp_h][tmp_w] = 0; // 이제 방문했으니 0으로 바꿔주고
							q.push({tmp_h,tmp_w}); // 다음 탐색을 위해 queue에 추가
						}
					}
				}
			}
		}
		
		cout << num << '\n';
	}
	
	return 0;
}

File: Python/4963_섬의개수.py
if __name__ == '__main__':
	dir = [[1,0],[-1,0],[0,1],[0,-1],[1,1],[1,-1],[-1,1],[-1,-1]] # 방향을 담는 배열(상,하,좌,우 대각선 4방)
	
	while True:
		w, h = list(map(int,input().split()))
		
		if w == 0 and h == 0: # 종료 조건
			break
		
		area = [] # 지도를 담은 배열
		for i in range(h):
			tmp = list(map(int,input().split()))
			area.append(tmp)
		
		num = 0
		for i in range(h):
			for j in range(w):
				if area[i][j] != 0: # 바다가 아닌 육지인 곳만 탐색(queue를 이용한 bfs)
					queue = []
					queue.append([i,j])
					area[i][j] = 0 # visit 배열 없이 방문 완료한 곳은 0으로 바꿔줌
					
					num += 1 # 하나의 root에 의해 탐색된 그래프는 하나의 영역이므로 영역 1개만 증가
					while queue: # bfs
						n = queue.pop(0)
						
						for d in dir:
							tmp_h = n[0]+d[0] 
							tmp_w = n[1]+d[1]
							if tmp_h >= 0 and tmp_h < h and tmp_w >= 0 and tmp_w < w: # 범위를 벗어나지 않고
								if area[tmp_h][tmp_w] != 0: # 바다가 아니라 육지일 때
									area[tmp_h][tmp_w] = 0 # 0으로 바꿔주고(여러 번 탐색하지 않도록)
									queue.append([tmp_h,tmp_w]) # 탐색을 위해 queue에 추가
									
		print(num)